### PR TITLE
Implementation of not type constraints

### PR DESCRIPTION
**Pull request by DakshChan**
This pr adds not type constraints to the minikanren implementation. Changes to common.rkt includes handling of not type constraints. Tests were added for not type constraints, and interactions with not type constraints.

### CODE REVIEW

File: common.rkt [hunk 1]
```diff
@@ -83,48 +88,83 @@
...
+              (and (eqv? type? u-type) st)
+              (state-simplify (state (state-sub st)
+                                     (state-diseq st)
+                                     (extend-types u type? (state-types st))
+                                     (state-distypes st)))))
+        (and (type? u) st))))
+
+;; Negation Constraints
+
+(define (extend-state/negated-diff-helper x newx acc)
```
Comment (xuexue) [inline common.rkt:106]: Use a better name here, something like "diff-prefix"

File: common.rkt [hunk 2]
```diff
@@ -83,48 +88,83 @@
...
-              (and (eqv? type? u-type) st)
-              (diseq-simplify (state (state-sub st)
-                                     (state-diseq st)
-                                     (extend-types u type? (state-types st))))))
-        (and (type? u) st))))
+(define (foldl/or proc acc lst)
+  (if (null? lst)
+      acc
+      (let ((new-acc (proc (car lst) acc)))
+        (or new-acc (foldl/and proc new-acc (cdr lst))))))
```
Comment (xuexue) [inline common.rkt:157]: Delete

File: common.rkt [hunk 3]
```diff
@@ -137,35 +177,42 @@
 (define (reified-index index)
   (string->symbol
     (string-append "_." (number->string index))))
+;(define (reify tm st) st)
```
Comment (xuexue) [inline common.rkt:180]: Delete

File: common.rkt [hunk 4]
```diff
@@ -137,35 +177,42 @@
...
-           (diseq (map pretty-diseq (filter-not contains-fresh? diseq)))
-           (diseq (map (lambda (=/=) (sort =/= term<?)) diseq))
-           (diseq (if (null? diseq) '() (list (cons '=/= diseq))))
-           (types (walk* (map pretty-types (state-types st)) results))
-           (types (filter-not contains-fresh? types))
-           (cxs (append types diseq)))
-      (if (null? cxs)
-          walked-sub
-          (Ans walked-sub (sort cxs term<?))))))
+                   (define t (walk tm (state-sub st))) ;; applies subsitiution
```
Comment (xuexue) [inline common.rkt:184]: Remove the comments

File: common.rkt [hunk 5]
```diff
@@ -190,16 +237,34 @@
     ((pair? v) 1)
     (else 1)))
 
+;; checks to see if x contains any fresh statments
```
Comment (xuexue) [inline common.rkt:240]: Remove this one

File: common.rkt [hunk 6]
```diff
@@ -190,16 +237,34 @@
...
 (define (contains-fresh? x)
   (if (pair? x)
-      (or (contains-fresh? (car x)) (contains-fresh? (cdr x)))
-      (var? x)))
-(define (pretty-diseq =/=s) (map (lambda (=/=) (let ((x (car =/=)) (y (cdr =/=))) (if (term<? x y) (list x y) (list y x)))) =/=s))
+      (or (contains-fresh? (car x)) (contains-fresh? (cdr x))) ;; if a pair, check both sides
+      (var? x))) ;; if no pair then: if it's variable we must have a fresh otherwise we don't
```
Comment (xuexue) [inline common.rkt:244]: Please remove comment

File: common.rkt [hunk 7]
```diff
@@ -190,16 +237,34 @@
...
+
+;; sorts by term comparison
+(define (pretty-diseq =/=s)
+  (map (lambda (=/=) (let ((x (car =/=)) (y (cdr =/=))) (if (term<? x y) (list x y) (list y x)))) =/=s))
+
+;; turns typed terms into pretty strings
 (define (pretty-types constraint) (list (type-check->sym (cdr constraint)) (car constraint)))
 
+;; returns string type for symbol, string, and number
```
Comment (xuexue) [inline common.rkt:253]: The ~~cake~~ comment is a lie

File: common.rkt [hunk 8]
```diff
@@ -190,16 +237,34 @@
...
+
+;; turns typed terms into pretty strings
+(define (pretty-distypes constraint) (list (distype-check->sym (cdr constraint)) (car constraint)))
+
+;; returns string type for symbol, string, and number
+(define (distype-check->sym pred)
+  (cond
+    ((eq? pred symbol?) '!sym)
+    ((eq? pred string?) '!str)
+    ((eq? pred number?) '!num)
```
Comment (xuexue) [inline common.rkt:269]: Instead of `!sym`, use `not-sym` due to Scheme/Racket conventions

File: microk-fo.rkt [hunk 9]
```diff
@@ -19,17 +22,20 @@
...
-(struct pause   (pause-state pause-goal) #:prefab)
+(struct disj     (g1 g2)                  #:prefab)
+(struct conj     (g1 g2)                  #:prefab)
+(struct relate   (thunk description)      #:prefab)
+(struct ==       (t1 t2)                  #:prefab)
+(struct =/=      (t1 t2)                  #:prefab)
+(struct symbolo  (t)                      #:prefab)
+(struct stringo  (t)                      #:prefab)
+(struct numbero  (t)                      #:prefab)
+(struct !symbolo (t)                      #:prefab)
```
Comment (xuexue) [inline microk-fo.rkt:33]: Use `not-` rather than `!`

File: common.rkt [hunk 10]
```diff
@@ -83,48 +88,83 @@
...
+         (newsub (and newst (state-sub newst)))
+         (newtypes (and newst (state-types newst))))
     (cond
       ((not newsub) st)
-      ((eq? newsub sub) #f)
-      (else (state sub (extend-diseq (disunify-helper sub newsub '()) diseq) types)))))
+      ((and (eq? mode 'sub) (not (eq? newsub sub)))
+       (state sub (extend-diseq (extend-state/negated-diff-helper sub newsub '()) diseq) types distypes)) 
+      ((and (eq? mode 'types) (not (eq? newtypes types)))
+       (state sub diseq types (extend-distypes (car (extend-state/negated-diff-helper types newtypes '())) distypes)))
```
Comment (xuexue) [inline common.rkt:123]:
```
(state sub 
        diseq
        types 
        (extend-distypes (car (extend-state/negated-diff-helper types newtypes '()))
        distypes)))
```

**Review by xuexue — CHANGES_REQUESTED** [review]

File: common.rkt (hunk 8 above)
Comment (DakshChan) [reply common.rkt]: https://github.com/DakshChan/first-order-miniKanren/commit/d5efb1eb9e2793c532e937f9bf75be51ffbea36f

File: microk-fo.rkt (hunk 9 above)
Comment (DakshChan) [reply microk-fo.rkt]: https://github.com/DakshChan/first-order-miniKanren/commit/d5efb1eb9e2793c532e937f9bf75be51ffbea36f

File: not-numbero-tests.rkt [hunk 11]
```diff
@@ -198,4 +198,12 @@
...
-  '(#s(Ans ((_.0 _.1)) ((=/= ((_.0 "wysi") (_.1 727))) (not-num _.0)))))
\ No newline at end of file
+  '(#s(Ans ((_.0 _.1)) ((=/= ((_.0 "wysi") (_.1 727))) (not-num _.0)))))
+
+(test 'not-numbero-13a
+  (run*(x) (not-numbero x) (conde ((numbero x)) ((== x "hello"))))
+  '(("hello")))
+
+(test 'not-numbero-13b
+  (run*(x) (not-numbero x) (conde ((== x 'x)) ((== x "hello"))))
```
Comment (xuexue) [inline not-numbero-tests.rkt:208]: Need a space here

File: distype-diseq-tests.rkt [hunk 12]
```diff
@@ -108,4 +108,49 @@
...
+            (conde ((not-stringo x)) ((not-numbero x)))
+            (conde ((=/= x "x")) ((=/= x 5))))
+      '(#s(Ans (_.0) ((not-str _.0)))
+        #s(Ans (_.0) ((=/= ((_.0 5))) (not-str _.0)))
+        #s(Ans (_.0) ((=/= ((_.0 "x"))) (not-num _.0)))
+        #s(Ans (_.0) ((not-num _.0)))))
+
+(test 'distype-diseq-22
+      (run* (x)
+            (conde ((not-stringo x)) ((not-numbero x)))
```
Comment (gregr) [inline distype-diseq-tests.rkt:131]: Add one more copy of this test, except replacing this `(conde ((not-stringo x)) ((not-numbero x)))` with `(conde ((not-stringo x)) ((not-numbero x)) ((not-symbolo x)))`

**Review by gregr — CHANGES_REQUESTED** [review]
I'd like to see some more tests where fresh variables are involved, so that we can exercise the reification's filtering of irrelevant variable constraints.

**Review by gregr — CHANGES_REQUESTED** [review]
Let's also have tests where, in addition to the constraints on individual variables, we have some interaction between the fresh and query variable, particularly `(=/= x y)`.

**Review by gregr — CHANGES_REQUESTED** [review]

File: distype-diseq-tests.rkt [hunk 13]
```diff
@@ -265,3 +265,55 @@
...
+            (=/= x y)
+            (stringo y))
+      '(#s(Ans (_.0 _.1) ((not-str _.0) (str _.1)))))
+
+(test 'distype-diseq-28f
+      (run* (x y)
+            (== x 5)
+            (stringo y)
+            (=/= x y))
+      '(#s(Ans (_.0 _.1) ((not-str _.0) (str _.1)))))
```
[the remark this answers is not in the document]
Comment (DakshChan) [reply distype-diseq-tests.rkt]: The tests fail. I think we made a copy-paste error. I fixed it in https://github.com/gregr/first-order-miniKanren/pull/3/commits/237d44de279f1061b8a0b72cca160b38cec5d1a8

**Comment by DakshChan** [comment]
> How have you been running the tests? The existing test script produces an error.
> 
> ```
> > ./test
> not-stringo-tests.rkt:5:13: not-stringo: unbound identifier
>   in: not-stringo
>   location...:
>    not-stringo-tests.rkt:5:13
> ```
> 
> You need to update the higher-order implementation too, not just the first-order one. Make sure both implementations pass all the tests.

Fixed this in the latest commit, all tests now pass for first and higher-order implementations

**Comment by gregr** [comment]
Thanks for working on this.

**Pull request merged**